observability/src/main/java/io/grpc/observability/Observability.java

```diff
@@ -16,49 +16,69 @@
 import io.grpc.observability.interceptors.InternalLoggingChannelInterceptor;
 import io.grpc.observability.interceptors.InternalLoggingServerInterceptor;
 import io.grpc.observability.logging.GcpLogSink;
 import io.grpc.observability.logging.Sink;
+import java.io.IOException;
 
 /** The main class for gRPC Observability features. */
 @ExperimentalApi("https://github.com/grpc/grpc-java/issues/8869")
 public final class Observability {
-  private static boolean initialized = false;
-  private static final String PROJECT_ID = "PROJECT";
+  private static Observability instance = null;
+  private final Sink sink;
 
   /**
    * Initialize grpc-observability.
    *
    * @throws ProviderNotFoundException if no underlying channel/server provider is available.
    */
-  public static synchronized void grpcInit() {
-    if (initialized) {
-      throw new IllegalStateException("Observability already initialized!");
+  public static synchronized Observability grpcInit() throws IOException {
+    if (instance == null) {
+      GlobalLoggingTags globalLoggingTags = new GlobalLoggingTags();
+      ObservabilityConfigImpl observabilityConfig = ObservabilityConfigImpl.getInstance();
+      Sink sink = new GcpLogSink(observabilityConfig.getDestinationProjectId());
+      instance = grpcInit(sink,
+          new InternalLoggingChannelInterceptor.FactoryImpl(sink,
+              globalLoggingTags.getLocationTags(), globalLoggingTags.getCustomTags(),
+              observabilityConfig),
+          new InternalLoggingServerInterceptor.FactoryImpl(sink,
+              globalLoggingTags.getLocationTags(), globalLoggingTags.getCustomTags(),
+              observabilityConfig));
+    }
+    return instance;
+  }
+
+  @VisibleForTesting static Observability grpcInit(Sink sink,
+      InternalLoggingChannelInterceptor.Factory channelInterceptorFactory,
+      InternalLoggingServerInterceptor.Factory serverInterceptorFactory) {
+    if (instance == null) {
+      instance = new Observability(sink, channelInterceptorFactory, serverInterceptorFactory);
     }
-    // TODO(dnvindhya): PROJECT_ID to be replaced with configured destinationProjectId
-    Sink sink = new GcpLogSink(PROJECT_ID);
-    LoggingChannelProvider.init(new InternalLoggingChannelInterceptor.FactoryImpl(sink));
-    LoggingServerProvider.init(new InternalLoggingServerInterceptor.FactoryImpl(sink));
-    // TODO(sanjaypujare): initialize customTags map
-    initialized = true;
+    return instance;
   }
 
-  /** Un-initialize or finish grpc-observability. */
-  // TODO(sanjaypujare): Once Observability is made into Singleton object,
-  //  close() on sink will be called as part of grpcFinish()
-  public static synchronized void grpcFinish() {
-    if (!initialized) {
-      throw new IllegalStateException("Observability not initialized!");
+  /** Un-initialize/shutdown grpc-observability. */
+  public synchronized void grpcShutdown() {
+    if (instance == null) {
```

Comment:
Thought of one potential issue. When you create the instance you are in a synchronized static method so you avoid any race conditions there. This method is synchronized as well, but since it is not static it is locking on the object, not the class instance. So I think there might be a chance for a race condition between grpInit() and grpcShutdown(). Not a likely race condition, but maybe we still want to make it impossible. Easiest fix I can think of is to lock on the class instance here.

Reply:
good catch. Made the change